ibft/storage/store.go

```diff
@@ -149,22 +150,35 @@ func (i *ibftStorage) CleanAllInstances(logger *zap.Logger, msgID []byte) error
 		return errors.Wrap(err, "failed to remove decided")
 	}
 
-	if err := i.delete(highestInstanceKey, msgID[:]); err != nil {
+	if err := i.delete(nil, highestInstanceKey, msgID[:]); err != nil {
 		return errors.Wrap(err, "failed to remove last decided")
 	}
 	return nil
 }
 
-func (i *ibftStorage) SaveParticipants(identifier convert.MessageID, slot phase0.Slot, operators []spectypes.OperatorID) error {
-	bytes, err := encodeOperators(operators)
+func (i *ibftStorage) UpdateParticipants(identifier convert.MessageID, slot phase0.Slot, newParticipants []spectypes.OperatorID) (updated bool, err error) {
+	txn := i.db.Begin()
+	defer txn.Discard()
+
+	existingParticipants, err := i.getParticipants(txn, identifier, slot)
 	if err != nil {
-		return err
+		return false, fmt.Errorf("get participants %w", err)
 	}
-	if err := i.save(bytes, participantsKey, identifier[:], uInt64ToByteSlice(uint64(slot))); err != nil {
-		return fmt.Errorf("could not save participants: %w", err)
+
+	mergedParticipants := mergeParticipants(existingParticipants, newParticipants)
+	if slices.Equal(mergedParticipants, existingParticipants) {
+		return false, nil
 	}
 
-	return nil
+	if err := i.saveParticipants(txn, identifier, slot, mergedParticipants); err != nil {
+		return false, fmt.Errorf("save participants: %w", err)
+	}
+
+	if err := txn.Commit(); err != nil {
+		return false, fmt.Errorf("commit transaction: %w", err)
+	}
+
+	return true, nil
 }
```

Comment:
im not sure what would happen if this func is called in parallel, is `i.db.Begin()` equivalent to a mutex lock in this case?

Reply:
Yeah, it seems BadgerDB transaction is not doing the kind of locking we typically expect from `transaction` as concept,

from what I see in their docs/code it locks only inside of `txn.Commit()` for a short period of time to check for conflicting writes (and returns error if there are any), I think they do it to get the most performance and provide optionality (because synchronous API everybody is used to isn't hard to build on top of that).

There is also some usage of transactions in `ssv/eth/eventhandler` package, but from cursory overview of the code it seems to only do sequential updates - so nothing to worry about.

Reply:
I thought it was sufficient without checking implementation details but actually I just looked at the implementation of `Commit` and now I think it's not enough, we need a mutex instead

Reply:
@iurii-ssv thanks for checking this
@nkryuchkov looks good now with mutex